File: solutions/javascript/secret-handshake/1/secret-handshake.js
const WINK         = 'wink';
const DOUBLE_BLINK = 'double blink';
const CLOSE_EYES   = 'close your eyes';
const JUMP         = 'jump';

const FLAGS = {
  wink:        1 << 0,
  doubleBlink: 1 << 1,
  closeEyes:   1 << 2,
  jump:        1 << 3,
  reverse:     1 << 4
};

export const commands = (bitmask) => {
  let actions = [];

  if (bitmask & FLAGS.wink)        actions.push(WINK);
  if (bitmask & FLAGS.doubleBlink) actions.push(DOUBLE_BLINK);
  if (bitmask & FLAGS.closeEyes)   actions.push(CLOSE_EYES);
  if (bitmask & FLAGS.jump)        actions.push(JUMP);

  return (
    bitmask & FLAGS.reverse
      ? actions.reverse()
      : actions
  );
};

File: solutions/ruby/secret-handshake/1/secret_handshake.rb
class SecretHandshake
  WINK              = 'wink'
  WINK_FLAG         = 1 << 0
  DOUBLE_BLINK      = 'double blink'
  DOUBLE_BLINK_FLAG = 1 << 1
  CLOSE_EYES        = 'close your eyes'
  CLOSE_EYES_FLAG   = 1 << 2
  JUMP              = 'jump'
  JUMP_FLAG         = 1 << 3
  REVERSE_FLAG      = 1 << 4

  def initialize(flags)
    @flags = flags
  end

  def commands
    actions = []

    return actions unless @flags.is_a? Integer

    actions << WINK         if @flags & WINK_FLAG         != 0
    actions << DOUBLE_BLINK if @flags & DOUBLE_BLINK_FLAG != 0
    actions << CLOSE_EYES   if @flags & CLOSE_EYES_FLAG   != 0
    actions << JUMP         if @flags & JUMP_FLAG         != 0

    return actions.reverse if @flags & REVERSE_FLAG != 0

    actions
  end
end
